### PR TITLE
[1.x] Extract hashtags

### LINKED ISSUE #109: [1.x] Hashtags are not extracted the right way
### Affected Versions

1.x

### Description

The hashtags extracted from the box description may contain symbols like `@` or `#` that are not part of the hashtag. This is because the regex used to extract the hashtags is not correct. I didn't test it with languages other than English, but I think it will also have problems with other languages.

**Acceptance Criteria:**

- [x] The hashtag should start with `#` and end with a space or a new line.
- [x] It should contain only letters, numbers, and underscores.
- [x] It should not contain symbols like `@` or `#`.
- [x] It should not contain emojis.
- [x] It should not contain URLs.
- [x] All characters should be in lowercase.
- [ ] It should work with all languages.
- [x] Isolate the hashtag extraction into standalone service
- [x] Write unit tests for the newly created service

### How to reproduce

```
curl --location --request POST 'https://localhost/api/v1/boxes' \
--header 'Authorization: Bearer {{accessToken}}' \
--form 'name="Consultant"' \
--form 'description="Repudiandae cupiditate dolorum veniam quisquam delectus harum nihil non.  #foo #bar #ff@bb #ff#bb"'
```

### Possible Solution

We can implement a separate service to extract the text's hashtags. This service can be injected into other services that need to extract hashtags from the text. 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by mgamal92**
| Q             | A                                                                  |
|---------------|--------------------------------------------------------------------|
| Branch?       | 1.x |
| Bug fix?      | yes  |
| New feature?  | no  |
| Deprecations? | no |
| Tickets       | Fix #109 |
| License       | MIT |

Note: I will create the separate service for that, now I'm working with POC

### CODE REVIEW

**Comment by mgamal92** [comment]
> @mgamal92 Any progress on this PR?

I'm so sorry, I was so busy the past days. I will give it more time this weekend

File: src/Foundation/Util/Hashtag.php [hunk 1]
```diff
@@ -16,9 +18,9 @@ final class Hashtag
     public function extract(string $text): array
     {
         $hashtags = [];
-        preg_match_all('/#(\S+)/', $text, $matches);
+        preg_match_all('/(^|\B)#(?![0-9_]+\b)([a-zA-Z0-9_]|\p{Arabic}){1,30}(\b|\r)/u', $text, $matches);
         foreach ($matches[0] as $match) {
-            $hashtags[] = mb_strtolower($match);
+            $hashtags[] = u(mb_strtolower($match))->toString();
```
Comment (imdhemy) [inline src/Foundation/Util/Hashtag.php:23]: Could you clarify the idea behind converting the match into a Unicode string?

**Review by imdhemy — CHANGES_REQUESTED** [review]
Add unit tests, please.

File: src/Foundation/Util/Hashtag.php (hunk 1 above)
Comment (mgamal92) [reply src/Foundation/Util/Hashtag.php]: It was an extra step so no need for that and removed.

**Review by imdhemy — CHANGES_REQUESTED** [review]
I added unit tests for different languages and some of them are broken.

https://github.com/piscibus/siklid-api/actions/runs/3728902225/jobs/6324311049#step:10:122

**Review by imdhemy — APPROVED** [review]
Great work! Finally, we have this excellent service! 🎉

**Comment by imdhemy** [comment]
@mgamal92 
Could you please submit the same PR in the following repo?

https://github.com/piscibus/php-hashtag

**Comment by mgamal92** [comment]
Why did you move it to separated repo ?

**Comment by imdhemy** [comment]
We already planned to publish a package for this service from day 1.

**Pull request merged**